### PR TITLE
chore(o3): make column memory size configurable

### PR DESCRIPTION
**Pull request by eugenels**
This change introduces fine-grained controls for overallocation when ingesting data. These are configuration options:
- Memory page size per column for O3 operations. Please be aware O3 will use 2x of this RAM per column
```
cairo.o3.column.memory.size=16M
```

- mmap sliding page size that TableWriter uses to append data for each column
```
cairo.writer.data.append.page.size=16M
```
- mmap page size for mapping small files, such as _txn, _todo and _meta. The default value is OS page size (4k Linux, 64K windows, 16k OSX M1). If you override this value it will be rounded to the nearest (greater) multiple of OS page size
```
cairo.writer.misc.append.page.size=4k
```
- mmap page size for appending index key data; key data is number of distinct symbol values times 4 bytes
```
cairo.writer.data.index.key.append.page.size=512k
```
- mmap page size for appending value data; value data are rowids, e.g. number of rows in partition times 8 bytes
 ```
cairo.writer.data.index.value.append.page.size=16M
```

## Example config for large number of small tables

```
cairo.o3.column.memory.size=256k
cairo.writer.data.append.page.size=256k
cairo.writer.data.index.key.append.page.size=64k
cairo.writer.data.index.value.append.page.size=256k
```

### CODE REVIEW

**Review by marregui — DISMISSED** [review]
LGTM

would it be worth adding an entry in `server.conf`:
```
#cairo.o3.column.memory.size=16777216
```

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|1|100.00%|

File: core/src/main/java/io/questdb/PropServerConfiguration.java [hunk 1]
```diff
@@ -661,6 +662,7 @@ public PropServerConfiguration(
             this.o3UpdPartitionSizeQueueCapacity = Numbers.ceilPow2(getInt(properties, env, "cairo.o3.upd.partition.size.queue.capacity", 128));
             this.o3PurgeDiscoveryQueueCapacity = Numbers.ceilPow2(getInt(properties, env, "cairo.o3.purge.discovery.queue.capacity", 128));
             this.o3PurgeQueueCapacity = Numbers.ceilPow2(getInt(properties, env, "cairo.o3.purge.queue.capacity", 128));
+            this.o3ColumnMemorySize = Numbers.ceilPow2(getIntSize(properties, env, "cairo.o3.column.memory.size", 16 * Numbers.SIZE_1MB));
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/PropServerConfiguration.java:665]: `ceilPow2` is incorrect. We have `Files.ceilPageSize()`

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

**Comment by bluestreak01** [comment]
we also need to configure append page size for table writer

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 10 / 10 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 10 / 10 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
Related to #1369

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 14 / 14 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 72 / 94 (76.60%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|11|90.91%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 74 / 96 (77.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 74 / 96 (77.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 74 / 96 (77.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 74 / 96 (77.08%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|11|12|91.67%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 73 / 91 (80.22%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|16|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 73 / 94 (77.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 73 / 94 (77.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 73 / 94 (77.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 73 / 94 (77.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/MemoryTag.java|0|19|00.00%|
|:large_blue_circle:|io/questdb/cairo/BitmapIndexWriter.java|23|25|92.00%|
|:large_blue_circle:|io/questdb/cairo/TableWriter.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/PropServerConfiguration.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/SymbolCache.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapReaderImpl.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/DefaultCairoConfiguration.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/vm/MemoryCMARWImpl.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/std/Files.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolColumnIndexer.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableBlockWriter.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cairo/SymbolMapWriter.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/AbstractIndexReader.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/TableReader.java|1|1|100.00%|

**Review by marregui — APPROVED** [review]
LGTM

**Pull request merged**